public/app/core/services/backend_srv.ts

```diff
@@ -195,59 +207,74 @@ export class BackendSrv implements BackendService {
-                canceler.resolve();
+    const cleanParams = omitBy(options.params, v => v === undefined || v.length === 0);
+    return fromFetch(options.params ? `${options.url}?${serializeParams(cleanParams)}` : options.url, {
+      method: options.method,
+      headers: {
+        'Content-Type': 'application/json',
+        Accept: 'application/json, text/plain, */*',
+        ...options.headers,
+      },
+      body: JSON.stringify(options.data),
+    })
+      .pipe(
+        mergeMap(res => res.json()),
+        tap(res => {
+          if (!options.silent) {
+            appEvents.emit(CoreEvents.dsRequestResponse, res);
+          }
+        }),
+        map(res => ({ data: res })),
+        retryWhen((attempts: Observable<any>) =>
+          attempts.pipe(
+            mergeMap((error, i) => {
+              const firstAttempt = i === 0;
+
+              if (error.status !== 401 || !requestIsLocal || !firstAttempt) {
+                return throwError(error);
               }
-              return this.datasourceRequest(options);
+
+              return from(this.loginPing());
             })
-            .catch((err: any) => {
-              if (err.status === 401) {
-                window.location.href = config.appSubUrl + '/logout';
-                throw err;
-              }
-            });
-        }
+          )
+        ),
+        catchError(err => {
+          if (err.status === this.HTTP_REQUEST_CANCELED) {
+            return throwError({ err, cancelled: true });
+          }
 
-        // populate error obj on Internal Error
-        if (_.isString(err.data) && err.status === 500) {
-          err.data = {
-            error: err.statusText,
-            response: err.data,
-          };
-        }
+          if (err.status === 401) {
+            window.location.href = config.appSubUrl + '/logout';
+            return throwError(err);
+          }
 
-        // for Prometheus
-        if (err.data && !err.data.message && _.isString(err.data.error)) {
-          err.data.message = err.data.error;
-        }
-        if (!options.silent) {
-          appEvents.emit(CoreEvents.dsRequestError, err);
-        }
-        throw err;
-      })
-      .finally(() => {
-        // clean up
-        if (options.requestId) {
-          this.inFlightRequests[options.requestId].shift();
-        }
-      });
+          // populate error obj on Internal Error
+          if (typeof err.data === 'string' && err.status === 500) {
+            err.data = {
+              error: err.statusText,
+              response: err.data,
+            };
+          }
+
+          // for Prometheus
+          if (err.data && !err.data.message && typeof err.data.error === 'string') {
+            err.data.message = err.data.error;
+          }
+
+          if (!options.silent) {
+            appEvents.emit(CoreEvents.dsRequestError, err);
+          }
+
+          return throwError(err);
+        }),
+        takeUntil(canceler ?? of()),
```

Comment:
Isn' this going to hang forever? I see the canceler only emits when consecutive requests wit the same ID are fired.